airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/report_streams.py

```diff
@@ -0,0 +1,302 @@
+    """
+    Custom exception occured when response with 429 status code received
+    """
+
+
+class ReportStream(BasicAmazonAdsStream, ABC):
+    """
+    Common base class for report streams
+    """
+
+    primary_key = None
+    CHECK_INTERVAL_SECONDS = 30
+    REPORT_WAIT_TIMEOUT = timedelta(hours=1)
+    REPORT_DATE_FORMAT = "%Y%m%d"
+    cursor_field = "reportDate"
+
+    def __init__(self, config: Spec, context: SourceContext, authenticator: Oauth2Authenticator):
+        self._authenticator = authenticator
+        self._session = requests.Session()
+        self._generate_model()
+        super().__init__(config, context)
+
+    @property
+    def model(self) -> JSModel:
+        return self._model
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+
+        if not stream_slice:
+            return
+        report_date = stream_slice[self.cursor_field]
+        report_infos = self._init_reports(report_date)
+        logger.info(f"Waiting for {len(report_infos)} report(s) to be generated")
+        start_time_point = datetime.now()
+        while report_infos and datetime.now() <= start_time_point + self.REPORT_WAIT_TIMEOUT:
+            completed_reports = []
+            logger.info(f"Checking report status, {len(report_infos)} report(s) remained")
+            for report_info in report_infos:
+                report_status, download_url = self._check_status(report_info)
+                if report_status == Status.FAILURE:
+                    logger.warn(f"Report for {report_info.profile_id} with {report_info.record_type} type generation failed")
+                    completed_reports.append(report_info)
+                elif report_status == Status.SUCCESS:
+                    metric_objects = self._download_report(report_info, download_url)
+                    for metric_object in metric_objects:
+                        yield self._model(
+                            profileId=report_info.profile_id,
+                            recordType=report_info.record_type,
+                            reportDate=report_date,
+                            metric=metric_object,
+                        ).dict()
+                    completed_reports.append(report_info)
+            for success_report in completed_reports:
+                report_infos.remove(success_report)
+            if report_infos:
+                logger.info(f"{len(report_infos)} report(s) remained, taking {self.CHECK_INTERVAL_SECONDS} seconds timeout")
+                time.sleep(self.CHECK_INTERVAL_SECONDS)
+        if not report_infos:
+            logger.info("All reports have been processed")
+        else:
+            logger.error("Not all reports has been processed due to timeout")
+
+        yield from []
+
+    def _generate_model(self):
```

Comment:
please add at least one line docstring, explaining why we doing this. This helps to review and read the code later

Reply:
done